atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/TransactionManagerTest.java

```diff
@@ -168,6 +170,74 @@ public void shouldNotConflictIfImmutableTimestampLockExpiresIfNoReadsOrWrites()
         txnManagerWithMocks.runTaskThrowOnConflict(txn -> null);
     }
 
+    @Test
+    public void writeDoesNotPopulateTimestampCache() {
+        Transaction transaction = txMgr.runTaskWithRetry(txn -> {
+            put(txn, TEST_TABLE, "row", "column", "test");
+            return txn;
+        });
+
+        assertThat(timestampCache.getCommitTimestampIfPresent(transaction.getTimestamp())).isNull();
+    }
+
+    @Test
+    public void readCachesWritersStartToCommitTimestamp() {
+        Transaction writer = txMgr.runTaskWithRetry(txn -> {
+            put(txn, TEST_TABLE, "row", "column", "test");
+            return txn;
+        });
+
+        assertThat(timestampCache.getCommitTimestampIfPresent(writer.getTimestamp())).isNull();
+
+        txMgr.runTaskWithRetry(txn -> {
+            get(txn, TEST_TABLE, "row", "column");
+            return txn;
+        });
+
+        assertThat(timestampCache.getCommitTimestampIfPresent(writer.getTimestamp()))
+                .isNotNull()
+                .isEqualTo(transactionService.get(writer.getTimestamp()));
```

Comment:
In general throughout this class, is `isNotNull()` really needed? I imagine the `isEqualTo` check should be good enough. `Transaction#getTimestamp` returns a primitive `long`, so non-null.